libkbfs/md_ops.go

```diff
@@ -53,14 +53,274 @@ func (md *MDOpsStandard) convertVerifyingKeyError(ctx context.Context,
+		// be read.  Then try currRmd.data.TLFPrivateKey to decrypt
+		// the leaf.  If it doesn't work, fetch MDs until we find the
+		// next keygen, and continue the loop.
+
+		privKey := currRmd.data.TLFPrivateKey
+		if !currRmd.IsReadable() {
+			pmd, err := decryptMDPrivateData(
+				ctx, md.config.Codec(), md.config.Crypto(),
+				md.config.BlockCache(), md.config.BlockOps(),
+				md.config.KeyManager(), md.config.Mode(), uid,
+				currRmd.GetSerializedPrivateMetadata(), currRmd,
+				head.ReadOnlyRootMetadata, md.log)
+			if err != nil {
+				return nil, err
+			}
+			privKey = pmd.TLFPrivateKey
+		}
+		currKeyGen := currRmd.LatestKeyGeneration()
+		if privKey == (kbfscrypto.TLFPrivateKey{}) {
+			return nil, errors.Errorf(
+				"Can't get TLF private key for key generation %d", currKeyGen)
+		}
+
+		mLeaf, err := eLeaf.Decrypt(
+			md.config.Codec(), privKey, kbfsRoot.Nonce, *kbfsRoot.EPubKey)
+		switch errors.Cause(err).(type) {
+		case nil:
+			return &mLeaf, nil
+		case libkb.DecryptionError:
+			// Fall-through to try another key generation.
+		default:
+			return nil, err
+		}
+
+		md.log.CDebugf(ctx, "Key generation %d didn't work; searching for "+
+			"the next one", currKeyGen)
+
+	fetchLoop:
+		for {
+			start := currRmd.Revision() + 1
+			end := start + maxMDsAtATime - 1 // range is inclusive
+			nextRMDs, err := getMergedMDUpdatesWithEnd(
+				ctx, md.config, currRmd.TlfID(), start, end, nil)
+			if err != nil {
+				return nil, err
+			}
+
+			for _, nextRmd := range nextRMDs {
+				if nextRmd.LatestKeyGeneration() > currKeyGen {
+					md.log.CDebugf(ctx, "Revision %d has key gen %d",
+						nextRmd.Revision(), nextRmd.LatestKeyGeneration())
+					currRmd = nextRmd.ReadOnlyRootMetadata
+					break fetchLoop
+				}
+			}
+
+			if len(nextRMDs) < maxMDsAtATime {
+				md.log.CDebugf(ctx,
+					"We tried all revisions and couldn't find a working keygen")
+				return nil, errors.Errorf("Can't decrypt merkle leaf")
+			}
+		}
+	}
+}
+
+func (md *MDOpsStandard) makeMerkleLeaf(
+	ctx context.Context, rmd ReadOnlyRootMetadata,
+	kbfsRoot *kbfsmd.MerkleRoot, leafBytes []byte) (
+	leaf *kbfsmd.MerkleLeaf, err error) {
+	if rmd.TlfID().Type() != tlf.Public {
+		return md.decryptMerkleLeaf(ctx, rmd, kbfsRoot, leafBytes)
+	}
+
+	var mLeaf kbfsmd.MerkleLeaf
+	err = md.config.Codec().Decode(leafBytes, &mLeaf)
+	if err != nil {
+		return nil, err
+	}
+	return &mLeaf, nil
+}
+
+func (md *MDOpsStandard) verifyKey(
+	ctx context.Context, rmds *RootMetadataSigned,
+	uid keybase1.UID, verifyingKey kbfscrypto.VerifyingKey,
+	irmd ImmutableRootMetadata) (cacheable bool, err error) {
+	err = md.config.KBPKI().HasVerifyingKey(ctx, uid, verifyingKey,
+		rmds.untrustedServerTimestamp)
+	var info revokedKeyInfo
+	switch e := errors.Cause(err).(type) {
+	case nil:
+		return true, nil
+	case RevokedDeviceVerificationError:
+		// TODO(KBFS-2963): Support team-keyed TLFs.
+		if irmd.TypeForKeying() == tlf.TeamKeying {
```

Comment:
FYI I added this check because I realized that the decryption totally doesn't work for team-backed TLFs, since the private key isn't stored in the private MD in that case but needs to be fetched from the service.  I made a new ticket for that and will have a separate PR for it.